lib/puma/cluster/fork_pipe.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+require "io/wait"
+
+module Puma
+  class Cluster
+    class ForkPipeReader
+      PAYLOAD_STRING = "q" # going with bigint out of learned fear of overflow
+      PAYLOAD_SIZE = 8 # # of bytes in a q payload
+
+      RESTART_SERVER = 0
+      START_REFORK = -1
+      AFTER_REFORK = -2
+
+      # avoids allocation of objects while reading
+      # only reads 1 payload at a time (minimize data loss if process exits unexpectedly)
+      # NOT thread-safe, but concurrent access not required anywhere currently
+      def initialize(pipe)
+        @pipe = pipe
+        @buffer = +""
+        @payload = +""
+      end
+
+      def read_nonblock
```

Comment:
the `_nonblock` methods are currently extraneous; I replaced them with simpler blocking methods below since changing to a simpler protocol (of letting the mold process handle when to call refork hooks) is a wider-sweeping change. I can delete them if preferred.